### PR TITLE
🎉 Merged my changes inside app.js - see discord

### PR DESCRIPTION
**Pull request by deqline**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,5 +1,13 @@
-// @ts-check
+//  _   _       _                       _          _    _       _       _               
+// | | | |     | |                     | |        | |  | |     | |     | |              
+// | | | | __ _| | ___  _ __ __ _ _ __ | |_ ______| |  | | __ _| |_ ___| |__   ___ _ __ 
+// | | | |/ _` | |/ _ \| '__/ _` | '_ \| __|______| |/\| |/ _` | __/ __| '_ \ / _ \ '__|
+// \ \_/ / (_| | | (_) | | | (_| | | | | |_       \  /\  / (_| | || (__| | | |  __/ |   
+//  \___/ \__,_|_|\___/|_|  \__,_|_| |_|\__|       \/  \/ \__,_|\__\___|_| |_|\___|_|   
+
```
Comment (PurpleMyst) [inline app.js:7]: I don't think we need the banner, we already have a README.md

File: app.js [hunk 2]
```diff
@@ -26,14 +34,15 @@ let run = true;
 const streamers = [];
 
 // ========================================== CONFIG SECTION =================================================================
-const CONFIG_PATH = "config.json";
+const CONFIG_PATH = "./config.json";
+let CONFIG_FILE = fs.existsSync(configPath) ? JSON.parse(fs.readFileSync(configPath, 'utf8')) : null;
```
Comment (PurpleMyst) [inline app.js:38]: I'd rather prefer we kept this in its own async function instead of doing I/O at the top level

File: app.js [hunk 3]
```diff
@@ -26,14 +34,15 @@ let run = true;
...
 const SCREENSHOT_FOLDER = "screenshots";
 const BASE_URL = "https://www.twitch.tv";
 
 const USER_AGENT =
   "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.108 Safari/537.36";
 
-const STREAMERS_URL = `${BASE_URL}/directory/game/VALORANT`;
+let STREAMERS_URL = `${BASE_URL}/directory/game/`;
```
Comment (PurpleMyst) [inline app.js:45]: Why'd you make this a `let`?

File: app.js [hunk 4]
```diff
@@ -78,9 +87,59 @@ const STREAM_WORST_QUALITY_QUERY =
...
+const DROP_INVENTORY_LIST = 'div.tw-flex-wrap.tw-tower.tw-tower--180.tw-tower--gutter-sm';
+const DROP_ITEM = '.tw-flex';
+const CATEGORY_NOT_FOUND = '[data-a-target="core-error-message"]';
+const DROP_STATUS = '[data-a-target="Drops Enabled"]';
 
 // ========================================== CONFIG SECTION =================================================================
 
+// ========================================== UTILS SECTION =================================================================
+/**
+* @fn idle
```
Comment (PurpleMyst) [inline app.js:99]: `@fn` is not needed if a JSDoc declaration precedes a function

File: app.js [hunk 5]
```diff
@@ -78,9 +87,59 @@ const STREAM_WORST_QUALITY_QUERY =
...
+* @param (ms) number of milliseconds to wait until the program resumes execution 
+* @author AlexSimpler
+* @return a promise which resolves after a timeout
```
Comment (PurpleMyst) [inline app.js:102]: You should add the return type

File: app.js [hunk 6]
```diff
@@ -78,9 +87,59 @@ const STREAM_WORST_QUALITY_QUERY =
...
+  return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+/**
+* @fn upperFirst
+* @param (word) makes the first letter of the word uppercase
+* @author AlexSimpler
+* @return the modified word
+*/
+function upperFirst(word) {
```
Comment (PurpleMyst) [inline app.js:114]: This is typically called "capitalize"

File: app.js [hunk 7]
```diff
@@ -78,9 +87,59 @@ const STREAM_WORST_QUALITY_QUERY =
...
 
+// ========================================== UTILS SECTION =================================================================
+/**
+* @fn idle
+* @param (ms) number of milliseconds to wait until the program resumes execution 
+* @author AlexSimpler
+* @return a promise which resolves after a timeout
+*/
+function idle(ms) {
+  return new Promise(resolve => setTimeout(resolve, ms));
```
Comment (PurpleMyst) [inline app.js:105]: What's the difference between this and `page.waitForTimeout`?

File: app.js [hunk 8]
```diff
@@ -78,9 +87,59 @@ const STREAM_WORST_QUALITY_QUERY =
...
+* @author AlexSimpler
+* @return the modified word
+*/
+function upperFirst(word) {
+  return (word[0].toUpperCase() + word.substring(1));
+}
+
+/**
+* @fn getUserProperty (Beta)
+* @param (page, name) takes as parameters the page handler aswell as the name of the property inside twilight-user
```
Comment (PurpleMyst) [inline app.js:120]: Separate these param declarations and give the type for them like in the other functions

File: app.js [hunk 9]
```diff
@@ -78,9 +87,59 @@ const STREAM_WORST_QUALITY_QUERY =
...
+* @author AlexSimpler
+* @return the value of the cookie
+*/
+async function getUserProperty(page, name) {
+
+  if (!name || !(/^[A-Za-z1-9]+$/.test(name))) throw new Error("Invalid cookie name: ", name);
+
+  const data = await page.cookies();
+  let cookieValue = undefined;
```
Comment (PurpleMyst) [inline app.js:129]: We can simplify this code by using `Array.prototype.find`

File: app.js [hunk 10]
```diff
@@ -78,9 +87,59 @@ const STREAM_WORST_QUALITY_QUERY =
...
+
+  for (let i = 0; i < data.length; i++) {
+    if (data[i].name == 'twilight-user') {
+      cookieValue = JSON.stringify((data[i].value).replace(/\%+[1-9]+/gm, ' ').replace(/\ \C\ /gm, ""));
+      cookieValue = cookieValue.replace(/"+/gm, "");
+      let reg = new RegExp(`(?<=${name}\\s\\:\\s)[a-zA-Z0-9]+`, 'gm');
+      cookieValue = cookieValue.match(reg);
+    }
+  }
+  return cookieValue ? cookieValue[0] : new Error("Invalid cookie returned");
```
Comment (PurpleMyst) [inline app.js:139]: Why do we return an error instead of throwing it?

File: app.js [hunk 11]
```diff
@@ -129,29 +188,59 @@ function header(message) {
  * @param {puppeteer.Browser} browser
  * @returns {Promise<boolean>} Are there any drops?
  */
-async function hasValorantDrop(browser) {
+async function hasDrops(browser) {
```
Comment (PurpleMyst) [inline app.js:191]: Parametrize this function by the game name

File: app.js [hunk 12]
```diff
@@ -129,29 +188,59 @@ function header(message) {
...
   const page = await browser.newPage();
   await page.goto(`${BASE_URL}/inventory`, { waitUntil: "networkidle2" });
   debug("Querying for drops ...");
-  const noDropsElement = await page.$(NO_DROPS_QUERY);
-  const noDrops = noDropsElement === null;
+  
+  /* Updated by AlexSimpler */
+  let noDrops = await query(page, NO_DROPS_QUERY);
+  //if there are noDrops then length > 0
+  if (noDrops.length) {
```
Comment (PurpleMyst) [inline app.js:201]: I dislike ints being used as booleans, make it more explicit by doing `noDrops.length !== 0`

File: app.js (hunk 11 above)
Comment (PurpleMyst) [inline app.js:191]: I'd prefer if we replaced the nested-ifs with guard-style ifs

File: app.js [hunk 13]
```diff
@@ -129,29 +188,59 @@ function header(message) {
...
+    
+    let count = 0;
+    let drop = await query(page, DROP_INVENTORY_LIST);
+    count = (await query(page, DROP_INVENTORY_LIST + ">" + DROP_ITEM)).length;
+    let received = false;
+
+    if (count) {
+      //just itterate through each notification and break if one drop has the name of the game name uppercase
+      for (let i = 0; i < count; i++) {
+        let game = (await query(page, `${DROP_INVENTORY_LIST + ">" + DROP_ITEM}:nth-child(${i + 1}) ${DROP_INVENTORY_NAME}`))
```
Comment (PurpleMyst) [inline app.js:216]: We can probably do this better with `page.$$eval` to avoid a lot of queries

File: app.js [hunk 14]
```diff
@@ -129,29 +188,59 @@ function header(message) {
...
+          .text().toUpperCase();
+        if (game == CONFIG_FILE.game.toUpperCase()) {
+          success = true;
+          break;
+        }
+      }
+      await idle(1500);
+      spinner.stop(1);
+      if (received) {
+        success(`🎉 Congrats you got ${(upperFirst(CONFIG_FILE.game))}!`);
```
Comment (PurpleMyst) [inline app.js:226]: We can skip the emoji, it doesn't show up right on windows

File: app.js [hunk 15]
```diff
@@ -237,24 +326,22 @@ async function watchRandomStreamers(browser, page) {
...
       await page.waitFor(jitter(1000));
       continue;
     }
 
     // Does the streamer have drops enabled?
-    const dropsEnabled = await page
-      .$eval('[data-a-target="Drops Enabled"]', () => true)
-      .catch(() => false);
+    //Updated by AlexSimpler
+    const dropsEnabled = const dropsEnabled = (await query(page, DROP_STATUS)).text();
```
Comment (PurpleMyst) [inline app.js:343]: I think this line is a typo?

File: app.js [hunk 16]
```diff
@@ -302,7 +389,7 @@ async function watchRandomStreamers(browser, page) {
 
     info(`Account status: ${status}`);
     info(`Time: ${dayjs().format("HH:mm:ss")}`);
-    info(`Watching stream for ${watchminutes} minutes`);
+    info(`Watching stream for ${watchminutes} minutes => ${dayjs().add((watchminutes), 'minutes').format('HH:mm:ss')}`);
```
Comment (PurpleMyst) [inline app.js:392]: I'd rather we used something like "ETA" instead of "=>"

File: app.js [hunk 17]
```diff
@@ -359,10 +446,18 @@ async function openBrowser() {
 async function getNewStreamers(page) {
   header("Streamer Refresh");
   await page.goto(STREAMERS_URL, { waitUntil: "networkidle0" });
+  
+  //was the category found?
+  const notFound = await query(page, CATEGORY_NOT_FOUND);
 
+  if (notFound.length || notFound.text() == "Category does not exist") {
+     error(`Game category not found, did you enter the game as displayed on twitch?`);
+     shutdown();
```
Comment (PurpleMyst) [inline app.js:455]: shutdown is a promise

File: app.js [hunk 18]
```diff
@@ -359,10 +446,18 @@ async function openBrowser() {
...
+  }
+  
   debug("Checking login ...");
   await checkLogin(page);
-
+
```
Comment (PurpleMyst) [inline app.js:460]: Run `prettier` on the code to remove whitespace issues like these

File: app.js [hunk 19]
```diff
@@ -380,10 +475,12 @@ async function getNewStreamers(page) {
  */
 async function checkLogin(page) {
   const cookies = await page.cookies();
-  if (cookies.findIndex((cookie) => cookie.name === "twilight-user") !== -1) {
-    success("Login successful!");
+  if (cookies.findIndex((cookie) => cookie.name === "twilight-user") !== -1) {#
```
Comment (PurpleMyst) [inline app.js:478]: What's the hash at the end?

File: app.js [hunk 20]
```diff
@@ -436,10 +533,12 @@ async function restartBrowser(browser) {
 }
 
 async function main() {
-  const { exec, token } = await readConfig();
+  //added game - AlexSimpler
+  const { exec, token, game} = await readConfig();
   browserConfig.executablePath = exec;
   authTokenCookie.value = token;
-
+  STREAMERS_URL = (STREAMERS_URL + game.toUpperCase();
```
Comment (PurpleMyst) [inline app.js:540]: UPPER_SNAKE_CASE means it's a costant, we should parametrize the function like above

File: app.js [hunk 21]
```diff
@@ -237,24 +326,22 @@ async function watchRandomStreamers(browser, page) {
     }
 
     // Is this streamer still streaming?
-    const channelStatus = await page
-      .$eval(CHANNEL_STATUS_QUERY, (el) => el.textContent)
-      .catch(() => "");
+    const channelStatus = (await query(page, CHANNEL_STATUS_QUERY)).text().trim().toUpperCase();
```
Comment (PurpleMyst) [inline app.js:329]: Does this line work? I removed `query()`

File: app.js [hunk 22]
```diff
@@ -237,24 +326,22 @@ async function watchRandomStreamers(browser, page) {
...
     info("Channel status: " + channelStatus);
     // We use `startsWith` because sometimes we get LIVELIVE
     // Also `toUpperCase` because sometimes we get LiveLIVE
     // This is because there are two elements with that class name
     // One below the player and one "in" the player
-    if (!channelStatus.toUpperCase().startsWith("LIVE")) {
+    if (!channelStatus.includes("LIVE")) {
```
Comment (PurpleMyst) [inline app.js:335]: We still need the `toUpperCase` due to the comment that I have above

**Review by PurpleMyst — CHANGES_REQUESTED** [review]
Made a few comments on the code

Some things that I only marked in one place actually need to be changed in multiple places (e.g. the JSDoc comments or the number-as-boolean in the `if`s)

See if you can make these changes :)

File: app.js (hunk 1 above)
Comment (deqline) [reply app.js:7]: Ok, you can remove it then.

File: app.js (hunk 2 above)
Comment (deqline) [reply app.js:38]: We can't because I am actually using it everywhere to then print the game name.

File: app.js (hunk 3 above)
Comment (deqline) [reply app.js:45]: because I am adding the provided game name in main from the `CONFIG_FILE`.

File: app.js (hunk 4 above)
Comment (deqline) [reply app.js:99]: Sure, you can remove it.

File: app.js (hunk 5 above)
Comment (deqline) [reply app.js:102]: yes, good point could you do that?

File: app.js (hunk 7 above)
Comment (deqline) [reply app.js:105]: None really  but I liked the esthetics, simpler. You could just change it to `page.waitForTimeout`

File: app.js (hunk 6 above)
Comment (AlexSimpler) [reply app.js:114]: Yeah, call it this way then

File: app.js (hunk 8 above)
Comment (deqline) [reply app.js:120]: Ok, well how would I do that, do i need to re-fork?

Comment (PurpleMyst) [reply app.js:120]: no you can make another commit

File: app.js (hunk 9 above)
Comment (AlexSimpler) [reply app.js:129]: Ok, good point, you can simplify it.

File: app.js (hunk 10 above)
Comment (deqline) [reply app.js:139]: Because i can't throw in a return statement.

File: app.js (hunk 2 above)
Comment (PurpleMyst) [reply app.js:38]: We can pass in the game names as a parameter, I dislike non-constant globals

File: app.js (hunk 20 above)
Comment (AlexSimpler) [reply app.js:540]: just lowercase it then, why add complexity?

File: app.js (hunk 18 above)
Comment (deqline) [reply app.js:460]: yeah I directly edited it on github.

File: app.js (hunk 19 above)
Comment (AlexSimpler) [reply app.js:478]: don't know, you can remove it

File: app.js (hunk 17 above)
Comment (deqline) [reply app.js:455]: forgot it was. await it then

File: app.js (hunk 16 above)
Comment (deqline) [reply app.js:392]: Ok, well your choice then :D

File: app.js (hunk 15 above)
Comment (deqline) [reply app.js:343]: yes, my bad, some rather bad copying from me.

File: app.js (hunk 21 above)
Comment (deqline) [reply app.js:329]: yes it works, don't remove anything

File: app.js (hunk 22 above)
Comment (deqline) [reply app.js:335]: Check the reply above

File: app.js (hunk 14 above)
Comment (deqline) [reply app.js:226]: Ok

File: app.js (hunk 13 above)
Comment (deqline) [reply app.js:216]: I don't have any issues regarding the query function but you could do that

File: app.js (hunk 12 above)
Comment (deqline) [reply app.js:201]: I think its an efficient way to use them because on cpp ints = bools, but nevermind you can replace it

File: app.js (hunk 11 above)
Comment (deqline) [reply app.js:191]: ok

Comment (deqline) [reply app.js:191]: Ok, can you do that, I don't really know where you're refering?

File: app.js (hunk 2 above)
Comment (deqline) [reply app.js:38]: well ok, but how can I modify the push?

File: app.js (hunk 21 above)
Comment (PurpleMyst) [reply app.js:329]: I don't mean that I removed it in this line, I mean that I removed the `query()` function completely a few commits ago

File: app.js (hunk 22 above)
Comment (PurpleMyst) [reply app.js:335]: Ah I didn't see you'd added `toUpperCase` directly on the declaration, nice

File: app.js (hunk 8 above)
Comment (deqline) [reply app.js:120]: ok, let me do that

File: app.js (hunk 20 above)
Comment (PurpleMyst) [reply app.js:540]: I don't think changing a global is a good way to do it, if we added a parameter maybe later we can check for multiple game drops

File: app.js (hunk 19 above)
Comment (PurpleMyst) [reply app.js:478]:
```suggestion
  if (cookies.findIndex((cookie) => cookie.name === "twilight-user") !== -1) {
```

**Pull request closed without merging**